File: services/analytics-service/src/analytics/analytics.service.ts
import { randomUUID } from 'node:crypto';

import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

import { DatabaseService } from '../database/database.service';
import { RequestContextService } from '../observability/request-context.service';
import { AggregateDto, TrackEventDto, UsageQueryDto } from './dto/analytics.dto';
import { DashboardPayload, KpiSummary, UsageTimePoint } from './entities/analytics.entity';

export interface UsageRow {
  date: string;
  memories_created: string;
  agents_run: string;
  tokens_used: string;
}

@Injectable()
export class AnalyticsService {
  private readonly logger = new Logger(AnalyticsService.name);

  constructor(
    private readonly db: DatabaseService,
    private readonly config: ConfigService,
    private readonly requestContext: RequestContextService,
  ) {}

  async getUsage(query: UsageQueryDto): Promise<{ data: UsageTimePoint[] }> {
    const tenantId = query.tenantId ?? this.requestContext.tenantId ?? 'default';
    const dateFrom = query.dateFrom ?? new Date(Date.now() - 30 * 86400000).toISOString();
    const dateTo = query.dateTo ?? new Date().toISOString();
    const trunc = (query.interval ?? 'day') === 'month' ? 'month' : (query.interval ?? 'day') === 'week' ? 'week' : 'day';

    const sql = `
      SELECT
        DATE_TRUNC($1, d)::date AS date,
        COALESCE(SUM(u.memories_created), 0) AS memories_created,
        COALESCE(SUM(u.agents_run), 0) AS agents_run,
        COALESCE(SUM(u.tokens_used), 0) AS tokens_used
      FROM generate_series($2::timestamptz, $3::timestamptz, '1 day') AS d
      LEFT JOIN usage_records u ON DATE_TRUNC($1, u.day)::date = d::date AND u.tenant_id = $4
      GROUP BY 1
      ORDER BY 1 ASC`;

    const result = await this.db.query<UsageRow>(sql, [trunc, dateFrom, dateTo, tenantId]);

    return {
      data: result.rows.map((r) => ({
        date: String(r.date),
        memoriesCreated: Number(r.memories_created),
        agentsRun: Number(r.agents_run),
        tokensUsed: Number(r.tokens_used),
      })),
    };
  }

  async getMetrics(): Promise<KpiSummary> {
    const tenantId = this.requestContext.tenantId ?? 'default';

    const [mem, agent, users, resp] = await Promise.all([
      this.db.query<{ count: string }>(
        `SELECT COUNT(*) AS count FROM memories WHERE tenant_id = $1 AND status != 'DELETED'`,
        [tenantId],
      ),
      this.db.query<{ count: string }>(`SELECT COUNT(*) AS count FROM agents WHERE tenant_id = $1`, [tenantId]),
      this.db.query<{ count: string }>(
        `SELECT COUNT(DISTINCT user_id) AS count FROM agent_executions WHERE tenant_id = $1 AND user_id IS NOT NULL`,
        [tenantId],
      ),
      this.db.query<{ avg: string | null }>(
        `SELECT AVG(duration_ms) AS avg FROM agent_executions WHERE tenant_id = $1 AND duration_ms IS NOT NULL`,
        [tenantId],
      ),
    ]);

    return {
      totalMemories: Number(mem.rows[0]!.count),
      totalAgents: Number(agent.rows[0]!.count),
      activeUsers: Number(users.rows[0]!.count),
      avgResponseTimeMs: resp.rows[0]!.avg ? Math.round(Number(resp.rows[0]!.avg)) : 0,
    };
  }

  async trackEvent(dto: TrackEventDto): Promise<{ id: string; message: string }> {
    const tenantId = dto.tenantId ?? this.requestContext.tenantId ?? 'default';
    const userId = this.requestContext.userId ?? null;
    const id = randomUUID();

    const result = await this.db.query(
      `INSERT INTO analytics_events (id, name, properties, tenant_id, user_id)
       VALUES ($1, $2, $3::jsonb, $4, $5) RETURNING id`,
      [id, dto.name, JSON.stringify(dto.properties ?? {}), tenantId, userId],
    );

    if (result.rows.length === 0) {
      throw new Error('Failed to insert analytics event');
    }

    this.logger.log({ id, name: dto.name }, 'Analytics event tracked');
    return { id, message: 'Event tracked' };
  }

  async getDashboard(): Promise<DashboardPayload> {
    const [kpis, usage] = await Promise.all([this.getMetrics(), this.getUsage({})]);

    return {
      kpis,
      usage: usage.data,
      generatedAt: new Date(),
    };
  }

  async aggregate(dto: AggregateDto): Promise<{ date: string; recordsCreated: number; message: string }> {
    const tenantId = dto.tenantId ?? this.requestContext.tenantId ?? 'default';
    const day = dto.date ?? new Date(Date.now() - 86400000).toISOString().slice(0, 10);

    const memResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) AS count FROM memories WHERE tenant_id = $1 AND DATE(created_at) = $2`,
      [tenantId, day],
    );
    const execResult = await this.db.query<{ count: string }>(
      `SELECT COUNT(*) AS count FROM agent_executions WHERE tenant_id = $1 AND DATE(created_at) = $2`,
      [tenantId, day],
    );
    const tokResult = await this.db.query<{ sum: string | null }>(
      `SELECT COALESCE(SUM(tokens_used), 0) AS sum FROM usage_records WHERE tenant_id = $1 AND day = $2`,
      [tenantId, day],
    );

    const result = await this.db.query(
      `INSERT INTO usage_records (id, day, tenant_id, memories_created, agents_run, tokens_used)
       VALUES ($1, $2, $3, $4, $5, $6)
       ON CONFLICT (tenant_id, day) DO UPDATE
         SET memories_created = EXCLUDED.memories_created,
             agents_run = EXCLUDED.agents_run,
             tokens_used = EXCLUDED.tokens_used
       RETURNING id`,
      [
        randomUUID(),
        day,
        tenantId,
        Number(memResult.rows[0]!.count),
        Number(execResult.rows[0]!.count),
        Number(tokResult.rows[0]!.sum ?? 0),
      ],
    );

    return {
      date: day,
      recordsCreated: result.rowCount ?? 0,
      message: 'Daily aggregation completed',
    };
  }
}

File: apps/backend/src/backend/services/analytics_service.py
import json
import uuid
from datetime import datetime, timezone

from sqlalchemy import text

from ..schemas.analytics import KpiSummary, UsageTimePoint


class AnalyticsService:
    async def get_usage(
        self,
        tenant_id: str,
        date_from: str | None,
        date_to: str | None,
        interval: str,
        db=None,
    ) -> list[UsageTimePoint]:
        params: dict = {}
        date_from = date_from or (datetime.now(timezone.utc).isoformat()[:10])
        date_to = date_to or (datetime.now(timezone.utc).isoformat()[:10])

        sql = """
            WITH date_series AS (
                SELECT generate_series(
                    :date_from::date,
                    :date_to::date,
                    :interval::interval
                ) AS bucket
            )
            SELECT
                to_char(ds.bucket, 'YYYY-MM-DD') AS date,
                COALESCE(SUM(ur.memories_created), 0) AS memories_created,
                COALESCE(SUM(ur.agents_run), 0) AS agents_run,
                COALESCE(SUM(ur.tokens_used), 0) AS tokens_used
            FROM date_series ds
            LEFT JOIN usage_records ur
                ON DATE_TRUNC(:trunc, ur.recorded_at) = ds.bucket
                AND ur.tenant_id = :tenant_id
            GROUP BY ds.bucket
            ORDER BY ds.bucket
        """

        trunc_map = {"day": "day", "hour": "hour", "week": "week", "month": "month"}
        trunc = trunc_map.get(interval, "day")

        params["date_from"] = date_from
        params["date_to"] = date_to
        params["interval"] = f"1 {interval}" if interval in ("day", "month") else "1 day"
        params["trunc"] = trunc
        params["tenant_id"] = tenant_id

        result = await db.execute(text(sql), params)
        rows = result.fetchall()

        return [UsageTimePoint(date=str(r[0]), memories_created=r[1], agents_run=r[2], tokens_used=r[3]) for r in rows]

    async def get_metrics(self, tenant_id: str, db=None) -> KpiSummary:
        import asyncio

        async def count_memories():
            r = await db.execute(
                text("SELECT COUNT(*) FROM memories WHERE tenant_id = :tenant_id"),
                {"tenant_id": tenant_id},
            )
            return r.scalar_one() or 0

        async def count_agents():
            r = await db.execute(
                text("SELECT COUNT(*) FROM agents WHERE tenant_id = :tenant_id"),
                {"tenant_id": tenant_id},
            )
            return r.scalar_one() or 0

        async def count_active_users():
            r = await db.execute(
                text("SELECT COUNT(DISTINCT user_id) FROM agent_executions WHERE tenant_id = :tenant_id"),
                {"tenant_id": tenant_id},
            )
            return r.scalar_one() or 0

        async def avg_response_time():
            r = await db.execute(
                text("SELECT COALESCE(AVG(response_time_ms), 0) FROM agent_executions WHERE tenant_id = :tenant_id"),
                {"tenant_id": tenant_id},
            )
            return float(r.scalar_one() or 0)

        total_memories, total_agents, active_users, avg_response_time_ms = await asyncio.gather(
            count_memories(), count_agents(), count_active_users(), avg_response_time(),
        )

        return KpiSummary(
            total_memories=total_memories,
            total_agents=total_agents,
            active_users=active_users,
            avg_response_time_ms=avg_response_time_ms,
        )

    async def track_event(
        self,
        name: str,
        properties: dict | None,
        tenant_id: str | None,
        user_id: str | None,
        db=None,
    ):
        event_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc)
        await db.execute(
            text("""
                INSERT INTO analytics_events (id, name, properties, tenant_id, user_id, created_at)
                VALUES (:id, :name, :properties, :tenant_id, :user_id, :created_at)
            """),
            {
                "id": event_id,
                "name": name,
                "properties": json.dumps(properties) if properties else None,
                "tenant_id": tenant_id,
                "user_id": user_id,
                "created_at": now,
            },
        )
        return event_id

    async def aggregate(self, date: str | None, tenant_id: str | None, db=None):
        target_date = date or datetime.now(timezone.utc).strftime("%Y-%m-%d")
        if tenant_id:
            memories_r = await db.execute(
                text("SELECT COUNT(*) FROM memories WHERE tenant_id = :tenant_id AND DATE(created_at) = :date"),
                {"tenant_id": tenant_id, "date": target_date},
            )
            agents_r = await db.execute(
                text("SELECT COUNT(*) FROM agents WHERE tenant_id = :tenant_id AND DATE(created_at) = :date"),
                {"tenant_id": tenant_id, "date": target_date},
            )
            tokens_r = await db.execute(
                text("""
                    SELECT COALESCE(SUM(total_tokens), 0) FROM agent_executions
                    WHERE tenant_id = :tenant_id AND DATE(created_at) = :date
                """),
                {"tenant_id": tenant_id, "date": target_date},
            )
            memories = memories_r.scalar_one() or 0
            agents = agents_r.scalar_one() or 0
            tokens = tokens_r.scalar_one() or 0

            await db.execute(
                text("""
                    INSERT INTO usage_records (id, tenant_id, recorded_at, memories_created, agents_run, tokens_used)
                    VALUES (:id, :tenant_id, :recorded_at, :memories, :agents, :tokens)
                    ON CONFLICT (tenant_id, recorded_at) DO UPDATE SET
                        memories_created = EXCLUDED.memories_created,
                        agents_run = EXCLUDED.agents_run,
                        tokens_used = EXCLUDED.tokens_used
                """),
                {
                    "id": str(uuid.uuid4()),
                    "tenant_id": tenant_id,
                    "recorded_at": target_date,
                    "memories": memories,
                    "agents": agents,
                    "tokens": tokens,
                },
            )
        else:
            await db.execute(
                text("""
                    INSERT INTO usage_records (id, tenant_id, recorded_at, memories_created, agents_run, tokens_used)
                    SELECT :id, tenant_id, :recorded_at,
                           COUNT(*) FILTER (WHERE DATE(created_at) = :recorded_at),
                           0, 0
                    FROM memories GROUP BY tenant_id
                """),
                {"id": str(uuid.uuid4()), "recorded_at": target_date},
            )


analytics_service = AnalyticsService()
